PXC-3265: Implement NBO OSU mode

dutow opened the pull request:
This commit implements non blocking updates for ALTER TABLE, CREATE
INDEX and DROP INDEX.

After this change, it'll be possible to set wsrep_osu_method to NBO for
a session, and execute these statements in non blocking mode.

Compared to TOI, NBO mode uses more relaxed synchronization: it takes
MDL locks and executes the transactions on all nodes, and then commits
them after all completes successfully.

dutow committed:
PXC-3265: Implement NBO OSU mode

This commit implements non blocking updates for ALTER TABLE, CREATE
INDEX and DROP INDEX.

After this change, it'll be possible to set wsrep_osu_method to NBO for
a session, and execute these statements in non blocking mode.

Compared to TOI, NBO mode uses more relaxed synchronization: it takes
MDL locks on all nodes using TOI, and then commits them after all
completes successfully using TOI again.

Conflict prevention doesn't only rely on MDL, it still uses the same
galera certification as TOI. This means that while it is possible to run
two parallel DDL statements, it is not possible to run two parallel DDLs
modifying the same table.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+
+    if (ret) {
+      assert(client_state.current_error());
+      // HERE's the cnonconflicting issue
```
```suggestion
      // HERE's the conflicting issue
```

dutow replied:
this was a debugging comment for me, removed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+    WSREP_DEBUG("%s", thd->wsrep_info);
+    thd_proc_info(thd, thd->wsrep_info);
+
+    /* DDL transaction are now atomic so append wsrep xid.
```
```suggestion
    /* DDL transactions are now atomic, so append wsrep xid.
```
OR
```suggestion
    /* DDL transaction is now atomic, so append wsrep xid.
```

dutow replied:
fixed (also in the TOI function, I copy pasted it from there)

venkatesh-prasad-v commented on sql/sql_show.cc:
```diff
@@ -2998,7 +2998,8 @@ class Fill_process_list : public Do_THD_Impl {
     ulonglong now_utime = my_micro_time();
 
-    if ((!inspect_thd->get_protocol()->connection_alive() &&
+    if ((inspect_thd->get_protocol() != nullptr &&
```
This is a PXC only change. So, I think this should be inside `#ifdef WITH_WSREP` block

dutow replied:
This change is no longer required, I fixed this issue since then, I reverted it.

venkatesh-prasad-v commented on sql/sql_show.cc:
```diff
@@ -4417,13 +4419,11 @@ static int get_schema_tmp_table_columns_record(THD *thd, TABLE_LIST *tables,
+        : field->is_flag_set(UNIQUE_KEY_FLAG)     ? "UNI"
+        : (field->is_flag_set(MULTIPLE_KEY_FLAG)) ? "MUL"
+        : (field->is_flag_set(CLUSTERING_FLAG))   ? "CLU"
+                                                  : "");
```
Is this change due to clang-format-10? 

I think it is better to have this code look like PS https://github.com/percona/percona-server/blob/release-8.0.26-16/sql/sql_show.cc#L4426-L4432.

dutow replied:
I accidentally ran clang-format-12 on the file, fixed.

venkatesh-prasad-v commented on sql/sql_table.cc:
```diff
@@ -17416,7 +17430,8 @@ bool mysql_alter_table(THD *thd, const char *new_db, const char *new_name,
     // either TOI or replicated and lock acquisition is a variant of NO_LOCK.
-    if ((wsrep_thd_is_toi(thd) || wsrep_thd_is_in_rsu(thd)) &&
+    if ((wsrep_thd_is_toi(thd) ||
+         wsrep_thd_is_in_rsu(thd) /* || wsrep_thd_is_in_nbo(thd) */) &&
```
This block was added while fixing PXC-3418, and I don't think this extensive locking is required for NBO. Please revert this change.

dutow replied:
removed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.transaction().ws_meta().seqno().get(),
+                WSREP_QUERY(thd));
+    // assert(strstr(WSREP_QUERY(thd), "f1") == 0);
```
Dead code?

dutow replied:
removed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+}
+
+static void wsrep_NBO_end(THD *thd) {
+  wsrep_to_isolation--;
```
Can we use pre-decrement here?
```suggestion
  --wsrep_to_isolation;
```

dutow replied:
fixed, and adjusted the TOI method similarly.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+    int ret = client_state.end_nbo_phase_two(err);
+
+    if (!ret) {
+      WSREP_DEBUG("TO END: %lld", client_state.nbo_meta().seqno().get());
```
```suggestion
      WSREP_DEBUG("NBO END: %lld", client_state.nbo_meta().seqno().get());
```

dutow replied:
fixed

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+        thd->get_transaction()->xid_state()->get_xid()->reset();
+      }
+    } else {
+      WSREP_WARN("TO isolation end failed for: %d, schema: %s, sql: %s", ret,
```
```suggestion
      WSREP_WARN("NBO end failed for: %d, schema: %s, sql: %s", ret,
```

dutow replied:
fixed

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+              ret, (thd->db().str ? thd->db().str : "(null)"),
+              WSREP_QUERY(thd));
+          if (!thd->is_error()) {
+            WSREP_WARN("Adjusting error to deadlock");
```
I don't think this message needs to be logged as a warning. Can we change it to
```suggestion
            WSREP_DEBUG("Adjusting error to deadlock");
```

dutow replied:
fixed, and removed the commented block below.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+      switch (client_state.current_error()) {
+        default:
+          WSREP_WARN(
+              "NBO isolation failed for: %d, schema: %s, sql: %s. "
```
Here and at other places. Do we need to mention the word `isolation` here?

dutow replied:
removed.

venkatesh-prasad-v commented on sql/wsrep_high_priority_service.cc:
```diff
@@ -636,11 +645,182 @@ int Wsrep_applier_service::apply_write_set(const wsrep::ws_meta &ws_meta,
+int Wsrep_applier_service::apply_nbo_begin(const wsrep::ws_meta &ws_meta,
+                                           const wsrep::const_buffer &data,
+                                           wsrep::mutable_buffer & /*err*/) {
+  DBUG_ENTER("Wsrep_applier_service::apply_nbo_begin");
```
`DBUG_ENTER` is not being used anywhere in MySQL. Can we use `DBUG_TRACE` instead?

dutow replied:
changed.

venkatesh-prasad-v commented on sql/wsrep_high_priority_service.cc:
```diff
@@ -636,11 +645,182 @@ int Wsrep_applier_service::apply_write_set(const wsrep::ws_meta &ws_meta,
+
+    assert(ret == 0);
+    if (ret != 0) {
+    }
```
Dead code?

dutow replied:
yes, I used it earlier for debugging, removed.

venkatesh-prasad-v commented on sql/wsrep_high_priority_service.cc:
```diff
@@ -636,11 +645,182 @@ int Wsrep_applier_service::apply_write_set(const wsrep::ws_meta &ws_meta,
+  snprintf(m_thd->wsrep_info, sizeof(m_thd->wsrep_info),
+           "wsrep: applying NBO write-set (%lld)",
+           (long long)wsrep_thd_trx_seqno(m_thd));
+  WSREP_DEBUG("%s", m_thd->wsrep_info);
```
Not an issue.

Here, the same message gets logged from both the parent and the child thread. Wouldn't this be redundant?
If this is required, then does it make sense to log it whether it is parent or child thread that is logging the message?

dutow replied:
Removed WSREP_DEBUG

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+  wsrep_to_isolation--;
+
+  wsrep::client_state &client_state(thd->wsrep_cs());
+  WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
+              WSREP_QUERY(thd));
```
Compilation fails here
```
[ 79%] Building CXX object sql/CMakeFiles/sql_main.dir/wsrep_var.cc.o
make[2]: *** [sql/CMakeFiles/sql_main.dir/build.make:141: sql/CMakeFiles/sql_main.dir/wsrep_high_priority_service.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
In file included from /home/venki/work/pxc/review/nbo/sql/sql_class.h:165,
                 from /home/venki/work/pxc/review/nbo/sql/rpl_commit_stage_manager.h:32,
                 from /home/venki/work/pxc/review/nbo/sql/binlog.h:52,
                 from /home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:16:
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc: In function ‘void wsrep_NBO_end(THD*)’:
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:2484:49: error: ‘class wsrep::client_state’ has no member named ‘nbo_meta’; did you mean ‘wsrep::ws_meta wsrep::client_state::nbo_meta_’? (not accessible from this context)
 2484 |   WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
      |                                                 ^~~~~~~~
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.h:283:25: note: in definition of macro ‘WSREP_LOG’
  283 |         .message(fmt, ##__VA_ARGS__);                              \
      |                         ^~~~~~~~~~~
/home/venki/work/pxc/review/nbo/sql/wsrep_mysqld.cc:2484:3: note: in expansion of macro ‘WSREP_DEBUG’
 2484 |   WSREP_DEBUG("NBO END: %lld: %s", client_state.nbo_meta().seqno().get(),
      |   ^~~~~~~~~~~
```
Can you check this?

dutow replied:
https://github.com/percona/wsrep-lib/pull/7 

Looks like I didn't add this function in my wsrep-lib change PR, but I had it in my local NBO checkout.

venkatesh-prasad-v replied:
Can you please merge the wsrep-lib PR and update this branch?

dutow replied:
done.

venkatesh-prasad-v reviewed (changes requested):
Hi Zsolt! Thank you for working on this. Here are my first round of review comments.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2494,20 +2488,19 @@ static void wsrep_NBO_end(THD *thd) {
       }
     } else {
-      WSREP_WARN("TO isolation end failed for: %d, schema: %s, sql: %s", ret,
+      WSREP_WARN("NBO isolation end failed for: %d, schema: %s, sql: %s", ret,
```
Remove the word `isolation` here.

venkatesh-prasad-v commented on mysql-test/suite/galera_ee/t/GCF-329A.test:
```diff
@@ -40,4 +40,4 @@ CALL mtr.add_suppression("Aborting");
 CALL mtr.add_suppression("Plugin 'InnoDB' will be forced to shutdown");
 CALL mtr.add_suppression("Slave SQL: Error 'Got error 7 during COMMIT'");
 CALL mtr.add_suppression("Adjusting error to deadlock");
```
This is no more a warning after it has been made to use `WSREP_DEBUG`. Can you please remove this suppression?

venkatesh-prasad-v commented on sql/sql_show.cc:
```diff
@@ -4345,8 +4344,7 @@ static int fill_temporary_tables(THD *thd, TABLE_LIST *tables, Item *cond) {
     if (store_temporary_table_record(thd, tables->table, tmp,
-                                     thd->lex->query_block->db,
-                                     thd->mem_root)) {
+                                     thd->lex->query_block->db, thd->mem_root)) {
```
Running clang-format-10 on this file produces a different output for this block. Can you check it?

```
diff --git a/sql/sql_show.cc b/sql/sql_show.cc
index c5a68592db4..65d7aabe8b4 100644
--- a/sql/sql_show.cc
+++ b/sql/sql_show.cc
@@ -4344,7 +4344,8 @@ static int fill_temporary_tables(THD *thd, TABLE_LIST *tables, Item *cond) {
 
   for (tmp = thd->temporary_tables; tmp; tmp = tmp->next) {
     if (store_temporary_table_record(thd, tables->table, tmp,
-                                     thd->lex->query_block->db, thd->mem_root)) {
+                                     thd->lex->query_block->db,
+                                     thd->mem_root)) {
       DBUG_RETURN(1);
     }
   }
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+                client_state.toi_meta().seqno().get(), WSREP_QUERY(thd));
+    WSREP_DEBUG("NBO END start: %lld: %s",
+                client_state.transaction().ws_meta().seqno().get(),
+                WSREP_QUERY(thd));
```
Looks that both these WSREP_DEBUG print same value.
```
2021-08-18T08:09:37.904049Z 17 [Note] [MY-000000] [WSREP] NBO END start: -1: alter table t1 engine=innodb
2021-08-18T08:09:37.904076Z 17 [Note] [MY-000000] [WSREP] NBO END start: -1: alter table t1 engine=innodb
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+
+  WSREP_DEBUG(
+      "Executing Query (%s) with write-set (%lld) and exec_mode: %s"
+      " in NBO Isolation mode",
```
Remove the word `Isolation` here.
```
2021-08-18T08:09:37.860245Z 17 [Note] [MY-000000] [WSREP] Executing Query (alter table t1 engine=innodb) with write-set (-1) and exec_mode: local in NBO Isolation mode
```

venkatesh-prasad-v commented on sql/sql_parse.h:
```diff
@@ -146,10 +146,17 @@ extern const LEX_CSTRING command_name[];
     wsrep_to_isolation_end(thd);
 
+#define WSREP_TO_ISOLATION_BEFORE_MDL_RELEASE \
+  if (WSREP(thd) && wsrep_nbo_isolation_end_begin(thd)) goto error;
```
`wsrep_nbo_isolation_end_begin()` is called unconditionally even in TOI mode. Can we check for `wsrep_thd_is_in_nbo()` before calling `wsrep_nbo_isolation_end_begin()`?

As of now, `NBO END maybe start` gets into the log for every command that is executed on the server with wsrep_debug enabled.
```
2021-08-18T08:13:04.534013Z 17 [Note] [MY-000000] [WSREP] Executing Query (alter table t1 engine=innodb) with write-set (-1) and exec_mode: local in TO Isolation mode
2021-08-18T08:13:04.534060Z 17 [Note] [MY-000000] [WSREP] wsrep: initiating TOI for write set (-1)
2021-08-18T08:13:04.534568Z 17 [Note] [MY-000000] [WSREP] Query (alter table t1 engine=innodb) with write-set (9) and exec_mode: toi replicated in TO Isolation mode
2021-08-18T08:13:04.534580Z 17 [Note] [MY-000000] [WSREP] wsrep: TO isolation initiated for write set (9)
2021-08-18T08:13:04.568648Z 17 [Note] [MY-000000] [WSREP] Updating WSREPXid: 2b3633ae-fffb-11eb-b6b0-efed34410649:9
2021-08-18T08:13:04.576848Z 17 [Note] [MY-000000] [WSREP] NBO END maybe start
2021-08-18T08:13:04.576945Z 17 [Note] [MY-000000] [WSREP] TO END: 9: alter table t1 engine=innodb

```

venkatesh-prasad-v commented on sql/sql_parse.h:
```diff
@@ -146,10 +146,17 @@ extern const LEX_CSTRING command_name[];
+#define WSREP_TO_ISOLATION_BEFORE_MDL_RELEASE \
+  if (WSREP(thd) && wsrep_nbo_isolation_end_begin(thd)) goto error;
+
+#define WSREP_TO_ISOLATION_AFTER_MDL_ACQUIRED \
```
Can we rename this to `WSREP_NBO_AFTER_MDL_ACQUIRED`?

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
   }
 }
 
+static int wsrep_NBO_begin(THD *thd, const char *db_, const char *table_,
```
We are having too many confusing functions like `wsrep_NBO_begin`,`wsrep_nbo_isolation_begin_end()`, `wsrep_nbo_isolation_end_begin()` and `wsrep_NBO_end`.

Can we try to avoid them and rename them to be aligned with `wsrep::client_state` member functions

Here is my suggestion:
```
+-------------------------------+---------------------------+
|            Existing           |          Proposed         |
+-------------------------------+---------------------------+
| wsrep_NBO_begin               | wsrep_NBO_begin_phase_one |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_begin_end | wsrep_NBO_end_phase_one   |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_end_begin | wsrep_NBO_begin_phase_two |
+-------------------------------+---------------------------+
| wsrep_NBO_end                 | wsrep_NBO_end_phase_two   |
+-------------------------------+---------------------------+
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+  return rc;
+}
+
+void wsrep_nbo_isolation_begin_end(THD *thd) {
```
We are having too many confusing functions like `wsrep_NBO_begin`,`wsrep_nbo_isolation_begin_end()`, `wsrep_nbo_isolation_end_begin()` and `wsrep_NBO_end`.

Can we try to avoid them and rename them to be aligned with `wsrep::client_state` member functions

Here is my suggestion:
```
+-------------------------------+---------------------------+
|            Existing           |          Proposed         |
+-------------------------------+---------------------------+
| wsrep_NBO_begin               | wsrep_NBO_begin_phase_one |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_begin_end | wsrep_NBO_end_phase_one   |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_end_begin | wsrep_NBO_begin_phase_two |
+-------------------------------+---------------------------+
| wsrep_NBO_end                 | wsrep_NBO_end_phase_two   |
+-------------------------------+---------------------------+
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+  }
+}
+
+int wsrep_nbo_isolation_end_begin(THD *thd) {
```
We are having too many confusing functions like `wsrep_NBO_begin`,`wsrep_nbo_isolation_begin_end()`, `wsrep_nbo_isolation_end_begin()` and `wsrep_NBO_end`.

Can we try to avoid them and rename them to be aligned with `wsrep::client_state` member functions

Here is my suggestion:
```
+-------------------------------+---------------------------+
|            Existing           |          Proposed         |
+-------------------------------+---------------------------+
| wsrep_NBO_begin               | wsrep_NBO_begin_phase_one |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_begin_end | wsrep_NBO_end_phase_one   |
+-------------------------------+---------------------------+
| wsrep_nbo_isolation_end_begin | wsrep_NBO_begin_phase_two |
+-------------------------------+---------------------------+
| wsrep_NBO_end                 | wsrep_NBO_end_phase_two   |
+-------------------------------+---------------------------+
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+}
+
+int wsrep_nbo_isolation_end_begin(THD *thd) {
+  WSREP_DEBUG("NBO END maybe start");
```
Can we instead print "NBO Phase 2 started" or something like that?

dutow replied:
I'll remove this line, it was only there for debugging purposes, we don't need it now.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,216 @@ static void wsrep_TOI_end(THD *thd) {
+
+  mysql_mutex_unlock(&thd->LOCK_thd_data);
+
+  if (is_thd_killed) {
```
Not related to this issue. But when the NBO query is killed in between the execution, it seems that the cluster enters into `Disconnected` state.

```
Session 1:
mysql> SET GLOBAL debug = 'd,sync.alter_locked_tables_inplace';
Query OK, 0 rows affected (0.00 sec)

Session 2:
mysql> show processlist;
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
| Id | User            | Host            | db   | Command | Time | State                    | Info                                    | Time_ms | Rows_sent | Rows_examined |
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
|  1 | system user     |                 | NULL | Sleep   | 7086 | wsrep aborter idle       | NULL                                    | 7086340 |         0 |             0 |
|  2 | system user     |                 | NULL | Sleep   | 6193 | innobase_commit_low (-1) | NULL                                    | 6192907 |         0 |             0 |
|  7 | event_scheduler | localhost       | NULL | Daemon  | 7084 | Waiting on empty queue   | NULL                                    | 7084327 |         0 |             0 |
| 18 | root            | localhost:38330 | test | Query   |    6 | debug sync point: now    | alter table t3 add column k varchar(10) |    6345 |         0 |             0 |
| 20 | root            | localhost:55980 | test | Query   |    0 | init                     | show processlist                        |       0 |         0 |             0 |
+----+-----------------+-----------------+------+---------+------+--------------------------+-----------------------------------------+---------+-----------+---------------+
5 rows in set (0.01 sec)

mysql> kill 18;
Query OK, 0 rows affected (0.00 sec)

Session 1:
mysql> alter table t3 add column l varchar(10);
ERROR 1047 (08S01): WSREP has not yet prepared node for application use

mysql> show status like 'wsrep_cluster_status%';
+----------------------+--------------+
| Variable_name        | Value        |
+----------------------+--------------+
| wsrep_cluster_status | Disconnected |
+----------------------+--------------+
1 row in set (0.00 sec)
```

venkatesh-prasad-v reviewed (changes requested):
Hi Zsolt! Thank you for addressing the review comments. 
While I'm still trying to understand the entire feature which needs some more time, I thought of publishing my second round of comments. Please find them below.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2296,223 @@ static void wsrep_TOI_end(THD *thd) {
+  if (buf) my_free(buf);
+
+  thd->wsrep_gtid_event_buf_len = 0;
+  thd->wsrep_gtid_event_buf = NULL;
```
Can we just call `free_gtid_event_buf(thd)` as `buf` will be same as `thd->wsrep_gtid_event_buf`? (See `wsrep_to_buf_helper` for more details)

dutow replied:
Fixed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+
+    WSREP_DEBUG(
+        "Query (%s) with write-set (%lld) and exec_mode: %s"
+        " replicated in TO Isolation mode",
```
```suggestion
        " replicated in NBO mode",
```

dutow replied:
Fixed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+        wsrep_thd_client_mode_str(thd));
+
+    snprintf(thd->wsrep_info, sizeof(thd->wsrep_info),
+             "wsrep: TO isolation initiated for write set (%lld)",
```
```suggestion
             "wsrep: NBO initiated for write set (%lld)",
```

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+int wsrep_NBO_begin_phase_two(THD *thd) {
+  wsrep::client_state &client_state(thd->wsrep_cs());
+  if (wsrep_thd_is_in_nbo(thd)) {
+    WSREP_DEBUG("NBO END start: %lld: %s",
```
```suggestion
    WSREP_DEBUG("NBO Phase Two start: %lld: %s",
```

dutow replied:
Fixed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+      switch (client_state.current_error()) {
+        default:
+          WSREP_WARN(
+              "NBO failed for: %d, schema: %s, sql: %s. "
```
```suggestion
              "NBO Phase Two failed for: %d, schema: %s, sql: %s. "
```

dutow replied:
Fixed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2593,7 +2880,6 @@ int wsrep_must_ignore_error(THD *thd) {
   const uint flags = sql_command_flags[thd->lex->sql_command];
 
   assert(error);
-  assert(wsrep_thd_is_toi(thd));
```
Instead of removing this assert, can we change the condition to 
```
assert(wsrep_thd_is_toi(thd) || wsrep_thd_is_in_nbo(thd));
```

dutow replied:
Fixed.

venkatesh-prasad-v commented on storage/innobase/trx/trx0sys.cc:
```diff
@@ -449,7 +449,7 @@ void trx_sys_update_wsrep_checkpoint(
         Allow current and new values be the same, without introducing new flags
         and logic to prevent double storing of the same value */
-        ut_ad(xid_seqno >= trx_sys_cur_xid_seqno);
+        //ut_ad(xid_seqno >= trx_sys_cur_xid_seqno);
```
Is this change necessarily required? This assertion has helped us in finding the cases which causes multiple requests to InnoDB to persist the xid_seqno. See https://jira.percona.com/browse/PXC-3442.

If it is possible, can we retain this assertion and fix the underlying issue?

dutow replied:
Fixed.

venkatesh-prasad-v commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+
+  bool is_thd_killed = thd->is_killed();
+
+  mysql_mutex_unlock(&thd->LOCK_thd_data);
+
+  thd->wsrep_skip_wsrep_hton = true;
```
In other places, `wsrep_skip_wsrep_hton` is set under the protection of `thd->LOCK_thd_data`.
See `wsrep_to_isolation_end()` in `sql/wsrep_mysqld.cc`
``` 
  2694   DEBUG_SYNC(thd, "wsrep_to_isolation_end_before_wsrep_skip_wsrep_hton");


  2695   mysql_mutex_lock(&thd->LOCK_thd_data);
  2696 
  2697   thd->wsrep_skip_wsrep_hton = false;
  2698 
  2699   mysql_mutex_unlock(&thd->LOCK_thd_data);
  2700   DEBUG_SYNC(thd, "wsrep_to_isolation_end_after_wsrep_skip_wsrep_hton");
```
```suggestion
  thd->wsrep_skip_wsrep_hton = true;
  mysql_mutex_unlock(&thd->LOCK_thd_data);
```

dutow replied:
Fixed.

venkatesh-prasad-v commented:
Also please extend the `galera_bf_bf*` testcases introduced as part of PXC-3449 with NBO for better coverage.

kamil-holubicki commented on sql/sql_alter.cc:
```diff
@@ -593,6 +594,8 @@ bool Sql_cmd_alter_table::execute(THD *thd) {
   /* This could be looked upon as too restrictive given it is taking
   a global mutex but anyway being TOI if there is alter tablespace
   operation active in parallel TOI would streamline it. */
+  // TODO: the above comment is no longer true with NBO, this could be
```
If not true anymore, why not update it?

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+              "Check wsrep connection state and retry the query.",
+              ret, (thd->db().str ? thd->db().str : "(null)"),
+              WSREP_QUERY(thd));
+          // NOT adjusting: for whatever reason is_error is unset,
```
dead code

dutow replied:
Removed.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
   }
 }
 
+static int wsrep_NBO_begin_phase_one(THD *thd, const char *db_,
```
This function and `wsrep_TOI_begin()` seem to have exactly the same logic with the difference that they this one calls another validator (`wsrep_can_run_in_nbo` instead of `wsrep_can_run_in_toi` - but both have the same prototype) and this one calls `begin_nbo_phase_one` instead of `enter_toi_local` - again the same prototype). 
Wouldn't it be better to have the common code only once and call the different parts via delegates?

dutow replied:
There are many small differeneces, I'm not sure if this would make it readable or easier to maintain.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2242,6 +2295,232 @@ static void wsrep_TOI_end(THD *thd) {
+              client_state.nbo_meta().seqno().get(), WSREP_QUERY(thd));
+
+  if (wsrep_thd_is_in_nbo(thd)) {
+    THD_STAGE_INFO(thd, stage_wsrep_completed_TO_isolation);
```
`wsrep_set_SE_checkpoint()` not needed? See TOI end.

dutow replied:
No, I had there originally, but it causes test failures.

kamil-holubicki commented on sql/wsrep_mysqld.cc:
```diff
@@ -2463,15 +2749,16 @@ bool wsrep_handle_mdl_conflict(const MDL_context *requestor_ctx,
 
   if (wsrep_thd_is_toi(request_thd) || wsrep_thd_is_in_rsu(request_thd) ||
-      wsrep_thd_is_applying(request_thd)) {
+      wsrep_thd_is_applying(request_thd) || wsrep_thd_is_in_nbo(request_thd)) {
```
Just  a side note: The logic in this  function becomes more and more obfuscated :(

kamil-holubicki commented on sql/sql_parse.h:
```diff
@@ -146,10 +146,17 @@ extern const LEX_CSTRING command_name[];
+  if (WSREP(thd) && wsrep_NBO_begin_phase_two(thd)) goto error;
+
+#define WSREP_NBO_AFTER_MDL_ACQUIRED \
+  if (wsrep_thd_is_local_nbo(thd)) wsrep_NBO_end_phase_one(thd);
```
Shouldn't these 2 macros be similar? I mean in 1st case we have check wif in NBO mode inside `wsrep_NBO_begin_phase_two()`, but for the 2nd one the check is directly in the macro - it looks like it could be moved inside `wsrep_NBO_end_phase_one()`. 
No need to check WSREP(thd) in 2nd macro?

kamil-holubicki commented on sql/sql_parse.h:
```diff
@@ -146,10 +146,17 @@ extern const LEX_CSTRING command_name[];
+      wsrep_thd_is_in_rsu(thd) || wsrep_thd_is_local_nbo(thd)) \
     wsrep_to_isolation_end(thd);
 
+#define WSREP_NBO_BEFORE_MDL_RELEASE \
```
I would name these macros to reflect what they really do, like
WSREP_NBO_2ND_PAHASE_BEGIN

kamil-holubicki reviewed (commented):
Hi @dutow , As we've been chatting on Slack, please remove test.cpp and wm.bak files that were accidentally added.
I'm fine with everything else.

kamil-holubicki reviewed (approved):
lgtm

venkatesh-prasad-v commented:
OK to squash. Please run Jenkins once and add the link as a comment.

venkatesh-prasad-v reviewed (approved):
Thank you for addressing all the review comments. Great work on this! Approved.

dutow committed:
PXC-3757: Alter table analyze partition crashes with NBO

dutow committed:
PXC-3757: Optimize table causes an assertion with NBO

dutow committed:
PXC-3265: fix NBO memory corruption

The data const_array might be freed and reused while the NBO
worker thread is still running, resulting in various crashes in
apply_events.

This commit copies the array before returning control to the applier
thread, preventing this issue.

dutow committed:
PXC-3756: fixing seqno assertions

dutow committed:
PXC-3756: Fixing another seqno assertion

dutow committed:
PXC-3265: fix missing binlog writes with NBO

There was a remaining todo and an additional condition in the wsrep
binlog checks to fix a crash during development. The underlying issue
was already fixed, but the condition remained there, keeping the applier
thread from writing NBO statements to the binlog.

Also modified the applier thread to enable/disable binlogging based on
the log_slave_updates variable.

dutow committed:
PXC-3757: ALTER TABLE TRUNCATE PARTITION crashes with NBO
